File: integrations/tool-tui/extensions/vscode/src/llmModels.ts
/**
 * LLM Model Definitions with Pricing
 * 
 * Comprehensive list of LLM models with their pricing per 1M tokens.
 * Data sourced from official provider pricing pages (January 2026).
 */

export interface LlmModel {
    name: string;
    provider: 'OpenAI/Azure' | 'Anthropic' | 'Google';
    contextWindow: string;  // e.g., "400K", "200K", "1M", "2M"
    inputPer1M: number;     // $ per 1M input tokens
    inputCachedPer1M: number; // $ per 1M cached input tokens
    outputPer1M: number;    // $ per 1M output tokens
}

/**
 * All supported LLM models with current pricing
 * Order: Anthropic (Claude) → Google (Gemini) → OpenAI
 */
export const LLM_MODELS: LlmModel[] = [
    // Anthropic Models (Claude)
    {
        name: 'Claude Opus 4.5',
        provider: 'Anthropic',
        contextWindow: '200K',
        inputPer1M: 5,
        inputCachedPer1M: 0.5,
        outputPer1M: 25,
    },
    {
        name: 'Claude Opus 4',
        provider: 'Anthropic',
        contextWindow: '200K',
        inputPer1M: 15,
        inputCachedPer1M: 1.5,
        outputPer1M: 75,
    },
    {
        name: 'Claude Sonnet 4.5',
        provider: 'Anthropic',
        contextWindow: '200K',
        inputPer1M: 3,
        inputCachedPer1M: 0.3,
        outputPer1M: 15,
    },
    {
        name: 'Claude Sonnet 4',
        provider: 'Anthropic',
        contextWindow: '200K',
        inputPer1M: 3,
        inputCachedPer1M: 0.3,
        outputPer1M: 15,
    },
    {
        name: 'Claude Haiku 4.5',
        provider: 'Anthropic',
        contextWindow: '200K',
        inputPer1M: 1,
        inputCachedPer1M: 0.1,
        outputPer1M: 5,
    },
    {
        name: 'Claude Haiku 3.5',
        provider: 'Anthropic',
        contextWindow: '200K',
        inputPer1M: 0.8,
        inputCachedPer1M: 0.08,
        outputPer1M: 4,
    },

    // Google Models (Gemini)
    {
        name: 'Gemini 3 Pro (Preview)',
        provider: 'Google',
        contextWindow: '1M',
        inputPer1M: 2,
        inputCachedPer1M: 0.2,
        outputPer1M: 12,
    },
    {
        name: 'Gemini 3 Flash (Preview)',
        provider: 'Google',
        contextWindow: '1M',
        inputPer1M: 0.5,
        inputCachedPer1M: 0.05,
        outputPer1M: 3,
    },
    {
        name: 'Gemini 2.5 Pro',
        provider: 'Google',
        contextWindow: '2M',
        inputPer1M: 1.25,
        inputCachedPer1M: 0.125,
        outputPer1M: 10,
    },
    {
        name: 'Gemini 2.5 Flash',
        provider: 'Google',
        contextWindow: '1M',
        inputPer1M: 0.3,
        inputCachedPer1M: 0.03,
        outputPer1M: 2.5,
    },
    {
        name: 'Gemini 2.5 Flash-Lite',
        provider: 'Google',
        contextWindow: '1M',
        inputPer1M: 0.1,
        inputCachedPer1M: 0.01,
        outputPer1M: 0.4,
    },
    {
        name: 'Gemini 2.0 Flash',
        provider: 'Google',
        contextWindow: '1M',
        inputPer1M: 0.1,
        inputCachedPer1M: 0.025,
        outputPer1M: 0.4,
    },
    {
        name: 'Gemini 2.0 Flash-Lite',
        provider: 'Google',
        contextWindow: '1M',
        inputPer1M: 0.075,
        inputCachedPer1M: 0,
        outputPer1M: 0.3,
    },

    // OpenAI/Azure Models
    {
        name: 'GPT-5.2',
        provider: 'OpenAI/Azure',
        contextWindow: '400K',
        inputPer1M: 1.75,
        inputCachedPer1M: 0.175,
        outputPer1M: 14,
    },
    {
        name: 'GPT-5.2 pro',
        provider: 'OpenAI/Azure',
        contextWindow: '400K',
        inputPer1M: 21,
        inputCachedPer1M: 0,
        outputPer1M: 168,
    },
    {
        name: 'GPT-5 mini',
        provider: 'OpenAI/Azure',
        contextWindow: '400K',
        inputPer1M: 0.25,
        inputCachedPer1M: 0.025,
        outputPer1M: 2,
    },
];

/**
 * Character-per-token ratios for different tokenizers (fallback)
 * These are empirically derived averages
 */
export const CHARS_PER_TOKEN: Record<string, number> = {
    'openai': 4.0,      // GPT models (cl100k_base, o200k_base)
    'anthropic': 3.8,   // Claude models
    'google': 4.2,      // Gemini models
};

/**
 * Get tokenizer type for a model
 */
export function getTokenizerType(model: LlmModel): 'openai' | 'anthropic' | 'google' {
    switch (model.provider) {
        case 'OpenAI/Azure': return 'openai';
        case 'Anthropic': return 'anthropic';
        case 'Google': return 'google';
    }
}

// Cache for token counts to avoid re-encoding
let tokenCache: Map<string, Map<string, number>> = new Map();
let lastText: string = '';

/**
 * Clear token cache when text changes
 */
function checkCacheValidity(text: string): void {
    if (text !== lastText) {
        tokenCache.clear();
        lastText = text;
    }
}

/**
 * Estimate token count for text using a specific model (sync fallback)
 */
export function estimateTokens(text: string, model: LlmModel): number {
    checkCacheValidity(text);
    
    const tokenizerType = getTokenizerType(model);
    
    // Check cache first
    if (tokenCache.has(tokenizerType)) {
        const cached = tokenCache.get(tokenizerType)!.get(model.name);
        if (cached !== undefined) return cached;
    }
    
    // Fallback to character-based estimation
    const charsPerToken = CHARS_PER_TOKEN[tokenizerType];
    return Math.max(1, Math.ceil(text.length / charsPerToken));
}

/**
 * Async token counting with real tiktoken (when available)
 */
export async function countTokensAsync(text: string, model: LlmModel): Promise<number> {
    checkCacheValidity(text);
    
    const tokenizerType = getTokenizerType(model);
    
    // Check cache
    if (!tokenCache.has(tokenizerType)) {
        tokenCache.set(tokenizerType, new Map());
    }
    const providerCache = tokenCache.get(tokenizerType)!;
    
    if (providerCache.has(model.name)) {
        return providerCache.get(model.name)!;
    }
    
    // Try to use tiktoken
    try {
        const { countTokens } = await import('./tiktoken');
        const result = await countTokens(text, tokenizerType);
        providerCache.set(model.name, result.tokens);
        return result.tokens;
    } catch {
        // Fallback to estimation
        const tokens = estimateTokens(text, model);
        providerCache.set(model.name, tokens);
        return tokens;
    }
}

/**
 * Calculate cost for a given token count
 */
export function calculateCost(
    tokens: number,
    model: LlmModel,
    type: 'input' | 'inputCached' | 'output'
): number {
    const pricePerMillion = type === 'input' 
        ? model.inputPer1M 
        : type === 'inputCached' 
            ? model.inputCachedPer1M 
            : model.outputPer1M;
    return (tokens / 1_000_000) * pricePerMillion;
}

/**
 * Format cost as string
 */
export function formatCost(cost: number): string {
    if (cost === 0) return '$0.0000';
    if (cost < 0.0001) return '<$0.0001';
    if (cost < 0.01) return `$${cost.toFixed(4)}`;
    if (cost < 1) return `$${cost.toFixed(4)}`;
    return `$${cost.toFixed(2)}`;
}

/**
 * Format token count with exact numbers
 */
export function formatTokenCount(tokens: number): string {
    return tokens.toLocaleString();
}

/**
 * Get models grouped by provider
 */
export function getModelsByProvider(): Map<string, LlmModel[]> {
    const grouped = new Map<string, LlmModel[]>();
    for (const model of LLM_MODELS) {
        const existing = grouped.get(model.provider) || [];
        existing.push(model);
        grouped.set(model.provider, existing);
    }
    return grouped;
}

File: integrations/tool-tui/crates/markdown/src/llm_models.rs
//! LLM Model Definitions with Pricing
//!
//! Comprehensive list of LLM models with their pricing per 1M tokens.
//! Data sourced from official provider pricing pages (January 2026).
//!
//! # Stability
//!
//! This module is **experimental** and not part of the stable API.
//! It may use `unwrap()` and `expect()` for convenience as it's not production code.

// Allow unwrap/expect in experimental LLM models code
#![allow(clippy::unwrap_used)]
#![allow(clippy::expect_used)]

use std::fmt;

/// LLM Provider
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Provider {
    OpenAI,
    Anthropic,
    Google,
}

impl fmt::Display for Provider {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Provider::OpenAI => write!(f, "OpenAI/Azure"),
            Provider::Anthropic => write!(f, "Anthropic"),
            Provider::Google => write!(f, "Google"),
        }
    }
}

/// LLM Model definition with pricing
#[derive(Debug, Clone)]
pub struct LlmModel {
    pub name: &'static str,
    pub provider: Provider,
    pub context_window: &'static str,
    pub input_per_1m: f64,
    pub input_cached_per_1m: f64,
    pub output_per_1m: f64,
    pub chars_per_token: f64,
}

/// All supported LLM models (January 2026)
pub static LLM_MODELS: &[LlmModel] = &[
    // OpenAI/Azure
    LlmModel {
        name: "GPT-5.2",
        provider: Provider::OpenAI,
        context_window: "400K",
        input_per_1m: 1.75,
        input_cached_per_1m: 0.175,
        output_per_1m: 14.0,
        chars_per_token: 4.0,
    },
    LlmModel {
        name: "GPT-5.2 pro",
        provider: Provider::OpenAI,
        context_window: "400K",
        input_per_1m: 21.0,
        input_cached_per_1m: 0.0,
        output_per_1m: 168.0,
        chars_per_token: 4.0,
    },
    LlmModel {
        name: "GPT-5 mini",
        provider: Provider::OpenAI,
        context_window: "400K",
        input_per_1m: 0.25,
        input_cached_per_1m: 0.025,
        output_per_1m: 2.0,
        chars_per_token: 4.0,
    },
    // Anthropic
    LlmModel {
        name: "Claude Opus 4.5",
        provider: Provider::Anthropic,
        context_window: "200K",
        input_per_1m: 5.0,
        input_cached_per_1m: 0.5,
        output_per_1m: 25.0,
        chars_per_token: 3.8,
    },
    LlmModel {
        name: "Claude Opus 4",
        provider: Provider::Anthropic,
        context_window: "200K",
        input_per_1m: 15.0,
        input_cached_per_1m: 1.5,
        output_per_1m: 75.0,
        chars_per_token: 3.8,
    },
    LlmModel {
        name: "Claude Sonnet 4.5",
        provider: Provider::Anthropic,
        context_window: "200K",
        input_per_1m: 3.0,
        input_cached_per_1m: 0.3,
        output_per_1m: 15.0,
        chars_per_token: 3.8,
    },
    LlmModel {
        name: "Claude Sonnet 4",
        provider: Provider::Anthropic,
        context_window: "200K",
        input_per_1m: 3.0,
        input_cached_per_1m: 0.3,
        output_per_1m: 15.0,
        chars_per_token: 3.8,
    },
    LlmModel {
        name: "Claude Haiku 4.5",
        provider: Provider::Anthropic,
        context_window: "200K",
        input_per_1m: 1.0,
        input_cached_per_1m: 0.1,
        output_per_1m: 5.0,
        chars_per_token: 3.8,
    },
    LlmModel {
        name: "Claude Haiku 3.5",
        provider: Provider::Anthropic,
        context_window: "200K",
        input_per_1m: 0.8,
        input_cached_per_1m: 0.08,
        output_per_1m: 4.0,
        chars_per_token: 3.8,
    },
    // Google
    LlmModel {
        name: "Gemini 3 Pro (Preview)",
        provider: Provider::Google,
        context_window: "1M",
        input_per_1m: 2.0,
        input_cached_per_1m: 0.2,
        output_per_1m: 12.0,
        chars_per_token: 4.2,
    },
    LlmModel {
        name: "Gemini 3 Flash (Preview)",
        provider: Provider::Google,
        context_window: "1M",
        input_per_1m: 0.5,
        input_cached_per_1m: 0.05,
        output_per_1m: 3.0,
        chars_per_token: 4.2,
    },
    LlmModel {
        name: "Gemini 2.5 Pro",
        provider: Provider::Google,
        context_window: "2M",
        input_per_1m: 1.25,
        input_cached_per_1m: 0.125,
        output_per_1m: 10.0,
        chars_per_token: 4.2,
    },
    LlmModel {
        name: "Gemini 2.5 Flash",
        provider: Provider::Google,
        context_window: "1M",
        input_per_1m: 0.3,
        input_cached_per_1m: 0.03,
        output_per_1m: 2.5,
        chars_per_token: 4.2,
    },
    LlmModel {
        name: "Gemini 2.5 Flash-Lite",
        provider: Provider::Google,
        context_window: "1M",
        input_per_1m: 0.1,
        input_cached_per_1m: 0.01,
        output_per_1m: 0.4,
        chars_per_token: 4.2,
    },
    LlmModel {
        name: "Gemini 2.0 Flash",
        provider: Provider::Google,
        context_window: "1M",
        input_per_1m: 0.1,
        input_cached_per_1m: 0.025,
        output_per_1m: 0.4,
        chars_per_token: 4.2,
    },
    LlmModel {
        name: "Gemini 2.0 Flash-Lite",
        provider: Provider::Google,
        context_window: "1M",
        input_per_1m: 0.075,
        input_cached_per_1m: 0.0,
        output_per_1m: 0.3,
        chars_per_token: 4.2,
    },
];

impl LlmModel {
    pub fn estimate_tokens(&self, text: &str) -> usize {
        ((text.chars().count() as f64) / self.chars_per_token).ceil() as usize
    }

    pub fn calculate_input_cost(&self, tokens: usize) -> f64 {
        (tokens as f64 / 1_000_000.0) * self.input_per_1m
    }

    pub fn calculate_cached_cost(&self, tokens: usize) -> f64 {
        (tokens as f64 / 1_000_000.0) * self.input_cached_per_1m
    }

    pub fn calculate_output_cost(&self, tokens: usize) -> f64 {
        (tokens as f64 / 1_000_000.0) * self.output_per_1m
    }
}

#[derive(Debug, Clone)]
pub struct TokenAnalysis {
    pub model_name: &'static str,
    pub provider: Provider,
    pub context_window: &'static str,
    pub tokens: usize,
    pub input_cost: f64,
    pub cached_cost: f64,
    pub output_cost: f64,
}

pub fn analyze_all_models(text: &str) -> Vec<TokenAnalysis> {
    LLM_MODELS
        .iter()
        .map(|m| {
            let tokens = m.estimate_tokens(text);
            TokenAnalysis {
                model_name: m.name,
                provider: m.provider,
                context_window: m.context_window,
                tokens,
                input_cost: m.calculate_input_cost(tokens),
                cached_cost: m.calculate_cached_cost(tokens),
                output_cost: m.calculate_output_cost(tokens),
            }
        })
        .collect()
}

pub fn format_cost(cost: f64) -> String {
    if cost == 0.0 {
        "$0.0000".into()
    } else if cost < 0.0001 {
        "<$0.0001".into()
    } else if cost < 1.0 {
        format!("${:.4}", cost)
    } else {
        format!("${:.2}", cost)
    }
}

pub fn format_tokens(tokens: usize) -> String {
    if tokens >= 1_000_000 {
        format!("{:.2}M", tokens as f64 / 1_000_000.0)
    } else if tokens >= 1_000 {
        format!("{:.1}K", tokens as f64 / 1_000.0)
    } else {
        tokens.to_string()
    }
}
